### ISSUE TITLE
Data augmentation

### ISSUE THREAD

**Issue by maxisme** [VisualComputingInstitute/triplet-reid#56]
Hello again,

I have just been looking through the project again and looking at ways to improve the accuracy of my model for facial recognition.

I am now looking at ways to generate more training data by performing some data augmentation. In your pipeline you [implement an (optional) random flip and crop](https://github.com/VisualComputingInstitute/triplet-reid/blob/f3aed745964d81d7410e1ebe32eb4329af886d2d/train.py#L262) to the dataset but should you not **append** this to the dataset opposed to **replacing**?

Also since adding a few more images (2,469,064 cropped faces of 9491 people) my model has become a lot harder to convolve.

I realise this is kind of a broad ML question and not really todo with this repo but would love to get your input!

Thank you

**Comment by Pandoro** [comment]
Hello @maxisme ,

I have thought of similar things too, but you have to realize what appending does to the model. Right now it gets a specific "input size" of persons and due to flipping or cropping, whatever the network sees does not actually change this size. At least this holds if the dataset is clean. However, if you append the data, you will need to shrink the original image since all images in a batch need the same size. This means that the network now has to learn on two scales, once it has to learn the original but cropped size and once it has to learn the shrinked size. While this should be possible, it now becomes a harder problem, especially since you don't have that much data for the later case since you don't perform crop augmentation on that size. If you want to go into this direction, I think a "zoom" augmentation would make more sense where you create a random crop that doesn't have a fixed size, but that could contain the full image, or only a crop with varying size. That way you will at least get a full range of scales, instead of just two discrete scales. All of this is obviously just an intuition and not a hard truth.

With flipping it shouldn't make a difference. In the long run you should see both version approximately equally often. The only difference you will see if you append it, is that your epochs become twice as long, so if you count epochs you should pay attention to this!

I hope that helps.

**Comment by lucasb-eyer** [comment]
It is standard practice to replace the image, note that this replacement is done afresh on every iteration. Appending it to the dataset could be done for flipping, it would double the dataset size, but for random crop, it is not feasible (memory-wise) to append all possible crops to the dataset. It is much better to simply re-sample a different random crop at each iteration.
